### ISSUE TITLE
Deepchem External Logger Integration (Wandb)

### ISSUE THREAD

**Issue by kshen3778** [deepchem/deepchem#2580]
Context: Following the discussion in the [June 25, 2021 developer call](https://docs.google.com/document/d/1bQ5aBgq4fn3e0rIhkZc3U-aVKfZN_0w-S2JzeEDt5hY/edit?ts=60d617aa) it was agreed to create a Github issue to discuss the direction for external loggers.

As I was working on the W&B logger integration I've found a few areas where more discussion and possibly some design changes may be needed. Although I'm working on the W&B integration, I feel like these areas would be useful to discuss for all future external logger integrations too. I've also added my own analysis for each of these areas. These are just my own thoughts and are just W&B specific.

Note: "Model class" will refer to the model classes where the W&B integration exists. Currently at the time of writing, it would be KerasModel and TorchModel (but could include JaxModel in the future too).

1. **Model checkpointing and versioning** ([Work in progress](https://github.com/deepchem/deepchem/pull/2577))

    For saving model checkpoints, the WandbLogger has been given a `save_model()` function which will handle saving and uploading checkpoint files to user's W&B account.

    There are two locations where checkpoints are being saved:

    - ValidationCallback: Calls `model.save_checkpoint` [here](https://github.com/deepchem/deepchem/blob/master/deepchem/models/callbacks.py#L88)
    - Model classes' `fit_generator()` (during training): KerasModel seems to be calling tensorflow's CheckpointManager, while TorchModel uses `save_checkpoint()` directly.
        - KerasModel: [https://github.com/deepchem/deepchem/blob/master/deepchem/models/keras_model.py#L462](https://github.com/deepchem/deepchem/blob/master/deepchem/models/keras_model.py#L462)
        - TorchModel: [https://github.com/deepchem/deepchem/blob/master/deepchem/models/torch_models/torch_model.py#L446](https://github.com/deepchem/deepchem/blob/master/deepchem/models/torch_models/torch_model.py#L446)

    Should WandbLogger.save_model function be:

    Option A: Called inside model classes' `save_checkpoint()`, perhaps somewhere at the [end](https://github.com/deepchem/deepchem/blob/master/deepchem/models/keras_model.py#L1072).

    or

    Option B: Called directly wherever model checkpointing is done along with `save_checkpoint()` rather than being called inside of it (an example of this is in the [current WIP PR](https://github.com/deepchem/deepchem/blob/2b7891f3d9b0737ec6012e611d041ef06f02c4ec/deepchem/models/callbacks.py#L92))

    Essentially, how much responsibility should be given to the existing `save_checkpoint()`? Is the model class's `save_checkpoint()` only responsible for the default Tensorflow/Torch checkpointing (option A) or should it call checkpointing for all loggers along with its default behaviour (option B).

    Furthermore, KerasModel's `fit_generator()` [uses tf.CheckpointManager directly](https://github.com/deepchem/deepchem/blob/master/deepchem/models/keras_model.py#L405) instead of calling `save_checkpoint()`. If this is the case, then only option B would make sense for KerasModel, since we can't stuff `WandbLogger.save_model()` into `tf.CheckpointManager`.

    Restoring from model checkpoint: 

    - A similar function called `restore_model()` can be added to WandbLogger.
    - Ideally there would be an option for restoring from Wandb in the model classes' `restore()` which is called by `load_from_pretrained()`. Is there another place that makes more sense?
    - Would it be appropriate to modify restore and `load_pretrained_model()` to support loading directly from Wandb? If so, what would be the best specify the new wandb `model_dir` rather than use the pre-existing `model_dir` argument.

        ```python
        # This could be an option, but what if there are multiple external loggers
        # each with their own specific restore code

        # Would they all need their own argument flag?

        model = dc.models.KerasModel(...)
        model.load_from_pretrained(...,wandb_model_dir="kshen/myproject/10pr4joa")
        ```

    - Another option would be to write a restore function inside WandbLogger and if someone wanted to use saved files they would just called WandbLogger's restore to download files to the local wandb run dir:

        ```python
        wandblogger = WandbLogger()
        wandblogger.restore_file(path_to_wandb_storage)
        model = dc.models.KerasModel(keras_model, wandb_logger=wandblogger)
        model.load_from_pretrained(model_dir=path_to_downloaded_model)
        model.fit(..)
        ```

        This would essentially move the functionality into the logger rather than keeping it in the model class's general logging API. It also doesn't seem very efficient to me as we are downloading a model and then passing it into `load_from_pretrained()`.

    - Once again, much like the dilemma with `save_checkpoint()` earlier, a central question seems to be how much responsibility should be given to a model classes' existing restore functions (`restore() and load_from_pretrained()`). Would they also be responsible for calling an external logger's restore code?
2. **Dataset saving and versioning**
    - Wandb supports saving/uploading a copy of the dataset at beginning of run for dataset versioning purposes.
    - Wandb would need access to the dataset in order to upload it
        - Dataset is passed into `fit()` but the model could also call fit_generator() directly
            - [https://github.com/deepchem/deepchem/pull/2520#issuecomment-856332764](https://github.com/deepchem/deepchem/pull/2520#issuecomment-856332764)
            - [https://github.com/deepchem/deepchem/pull/2520#discussion_r651274368](https://github.com/deepchem/deepchem/pull/2520#discussion_r651274368))
        - Would it still be appropriate for Wandb to access and save the dataset inside `fit()` only? Dataset saving will only happen if `fit()` is called - if the user directly calls `fit_generator()` then Wandb will not save any dataset since `fit()` is not called.

            ```python
            def fit(self,
                      dataset: Dataset,
                      nb_epoch: int = 10,
                      max_checkpoints_to_keep: int = 5,
                      checkpoint_interval: int = 1000,
                      deterministic: bool = False,
                      restore: bool = False,
                      variables: Optional[List[tf.Variable]] = None,
                      loss: Optional[LossFn] = None,
                      callbacks: Union[Callable, List[Callable]] = [],
                      all_losses: Optional[List[float]] = None) -> float:
                
            		#save dataset here
            		self.wandblogger.save_dataset(dataset)
            	
                return self.fit_generator(
                    self.default_generator(
                        dataset, epochs=nb_epoch,
                        deterministic=deterministic), max_checkpoints_to_keep,
                    checkpoint_interval, restore, variables, loss, callbacks, all_losses)
            ```

        - Another option is to pass in the dataset directly to WandbLogger as an argument when being initialized. However, this would mean dataset access/saving will be a logger functionality rather than being present in the model classes.
        - It depends if Deepchem would like access to dataset to be a feature of the general logger API, or something that is passed into each individual logger that needs access.
3. **Training logging**
    - It would be nice to log extra training metrics apart from the loss at that step: [https://github.com/deepchem/deepchem/pull/2520#issuecomment-855964524](https://github.com/deepchem/deepchem/pull/2520#issuecomment-855964524)
    - What would be the best way to calculate training metrics and log them (will need access to training dataset, and pass in the metrics to calculate)
        - Ideally it would be like calculating metrics on training dataset just like how its done in ValidationCallback [here](https://github.com/deepchem/deepchem/blob/master/deepchem/models/callbacks.py#L74).
        - The problem is that the user could call `fit_generator()` directly, which means there will be no dataset passed in to calculate metrics on.
            - [https://github.com/deepchem/deepchem/pull/2520#issuecomment-856332764](https://github.com/deepchem/deepchem/pull/2520#issuecomment-856332764)
            - [https://github.com/deepchem/deepchem/pull/2520#discussion_r651274368](https://github.com/deepchem/deepchem/pull/2520#discussion_r651274368))
    - Log epoch wise instead of step wise
        - No good way to get the epoch inside `default_generator()` or `fit_generator()`
        - Epoch could be calculated using the training dataset size and batch size, but this is contingent on the user calling `fit()` and the dataset not being created via generator.

These are all the areas I've looked at so far. I apologize for making this so long as I wanted to make sure it's detailed enough.

Would welcome discussion and feedback on how best to go about implementing these!

**Comment by peastman** [comment]
These are good questions.

For checkpointing, I think my inclination would be to keep it separate.  KerasModel provides features for loading and saving checkpoints with the standard TF mechanism.  TorchModel does with the standard PyTorch mechanism.  A logger could provide its own checkpointing mechanism, and a user might choose to use that mechanism instead of or in addition to the one provided by the model.  I would keep that internal to the logger.  All the model needs to know is whether it should write checkpoints itself.

Regarding epochs, what about this idea: we could add a new argument `batches_per_epoch=None` to `fit_generator()`.  If you call `fit()`, it would calculate the number of batches in each epoch and pass the correct value.  If you call `fit_generator()` directly, you would have the option of specifying that argument yourself.  If it's left as None but you request logging by epochs, it would throw an exception.  In that case, it doesn't have enough information to do what you've requested.

For calculating other metrics, let's distinguish between two cases: one that are calculated independently from the training process, and ones calculated as part of the training process.

- If you want to compute metrics on the training set at the end of each epoch, that's no different than calculating them on a validation set.  We can use the same mechanism for both.  The only difference is the choice of dataset.  But this is expensive, since it means an extra pass through the training set and evaluating it a second time, so...
- You might want to compute a metric based on the outputs from each batch that you're already computing during training.  To do this, we just need a mechanism to pass the inputs and outputs of each batch to the logger.

Thinking over all this and looking through the code, I'm starting to get an idea for a slightly different API.  It isn't fully worked out yet, but here's the general idea I have in mind.

First of all, let's create an abstract Logger class defining a generic logging API.  WandbLogger would subclass it.  The argument passed to the model constructor would just be `logger` instead of `wandb_logger`.  It would have the following methods.

- `setup(model)` called at the very end of the model's constructor.  (In your current implementation you also have a `update_config()` method that receives all the arguments passed to the model's constructor, but I don't think that's necessary.  If we just wait to the end of the constructor to call `setup()`, it can pull any information it needs directly from the model.)
- `finish()` called at the end of training.  (Keep in mind that it's fine to call `fit()` repeatedly, so `finish()` shouldn't prevent logging more information later.)
- `log_batch()` called after each training step.  It would receive all information a logger might plausibly want: the step index, the model inputs and outputs, the training loss, etc.
- `log_epoch()` called at the end of each epoch, if epochs have been defined.

Does that seem like it would work?

**Comment by kshen3778** [comment]
Thanks!

Checkpointing:
I think it would make sense to keep checkpointing specific to the logger as well. Would that mean the logger checkpointing functions would sexist in KerasModel/TorchModel `fit()` but outside of their `save_checkpoint()` functions?
In other words, is the methadology used in the current WIP PR for checkpointing in line with what you're thinking of?

Epoch logging:
Regarding epoch logging, a `batches_per_epoch=None` flag seems to make sense

Metric calcuation at training time:
I just want to make sure I understand correctly. Basically, if the metrics are calculated at the end of epoch, we could just use ValidationCallback since it's no different that just evaluating on the validation set right? Therefore, the only thing that would add value would be if we could calculate metrics for each training batch. Is my understanding correct?

Let me think about the last point about the Logger class. I'll add my thoughts later this week.

**Comment by kshen3778** [comment]
**Logger base class:**
I think those 4 methods you proposed seems good for now. I was thinking for more advanced functionalities such as model checkpointing, uploading files, the subclass (WandbLogger for example) would just have their own methods for example `save_model` and `restore_model`. Would that make sense?

Also as a side question, assuming we decide on the Logger base class, would this be written up and developed by the Deepchem team or would someone else need to implement it since you guys might not have the bandwidth.

**Comment by peastman** [comment]
> the subclass (WandbLogger for example) would just have their own methods for example save_model and restore_model. Would that make sense?

That sounds reasonable to me.

> would this be written up and developed by the Deepchem team

Yes, we'd take responsibility for maintaining the base class.  Which would be hardly any work at all, since all it would do is define a few methods without implementations. :)  But we might also decide to create some other logging implementations, and we would maintain those too.

**Comment by kshen3778** [comment]
Sounds good!

Any thoughts on the first section of points below especially for model checkpointing? Just want to make sure my understanding of keeping checkpointing separate is correct so I can proceed with the checkpointing PR.
> Thanks!
> 
> **Checkpointing:**
> I think it would make sense to keep checkpointing specific to the logger as well. Would that mean the logger checkpointing functions would still exist in KerasModel/TorchModel `fit()` but outside of their `save_checkpoint()` functions?
> In other words, is the methadology used in the current WIP PR for checkpointing in line with what you're thinking of?
> 
> Also, do you have any opinions regarding restoring models? If kept separate, I would assume we shouldn't touch the load_from_pretrained() and restore(). The user could download the model file from Wandb and then just pass it into load_from_pretrained.
> 
> **Epoch logging:**
> Regarding epoch logging, a `batches_per_epoch=None` flag seems to make sense.
> 
> **Metric calcuation at training time:**
> I just want to make sure I understand correctly. Basically, if the metrics are calculated at the end of epoch, we could just use ValidationCallback since it's no different that just evaluating on the validation set right? Therefore, the only thing that would add value would be if we could calculate metrics for each training batch. Is my understanding correct?
> 
> EDIT: I think it would also be useful for the ValidationCallback (VC) to take in multiple datasets. Or at least for fit() to support multiple ValidationCallbacks. The reason being if we go with the case of metrics being calculated independently from the training process, someone could very well have a VC for logging training metrics, and a VC for logging validation metrics. If that's the case, then there's no way for VC to differentiate between the different keys (since both training and validation will have the same key names). Therefore any logger will be logging the same key twice:
> https://github.com/deepchem/deepchem/blob/master/deepchem/models/callbacks.py#L81
> https://github.com/deepchem/deepchem/blob/master/deepchem/models/callbacks.py#L93
> Currently, I've made a [PR](https://github.com/deepchem/deepchem/pull/2586) for WandbLogger to go around this issue by differentiating different datasets being scored on inside the WandbLogger class, but it would be nice to see VC changed to take this into account.
> 
> Let me think about the last point about the Logger class. I'll add my thoughts later this week.

**Comment by peastman** [comment]
My thought is that checkpointing by the logger would be completely independent of anything done by KerasModel or TorchModel.  It would save checkpoints inside `log_batch()` or `log_epoch()`.  The model wouldn't know or care that the logger was doing its own checkpointing.

**Comment by kshen3778** [comment]
Thanks, that makes sense. One extra thing:
- The `logger` argument passed to the model constructor should be a list of loggers just in case the user wants to log to multiple logger (I guess if this is the case it should be renamed to `loggers`)

**Comment by kshen3778** [comment]
I was wondering if there should be a function similar to `log_batch()` that would be used in the ValidationCallback [logging](https://github.com/deepchem/deepchem/blob/master/deepchem/models/callbacks.py#L99): Perhaps called `log_score()`?

**Comment by peastman** [comment]
Let's keep the argument called `logger` and declare the type as `OneOrMany[Logger]`.  Usually people will just pass one value to it, but we'll also support a list if they want that.

> I was wondering if there should be a function similar to log_batch() that would be used in the ValidationCallback logging: Perhaps called log_score()?

That could make sense.  I think the function would be `log_values()`.  The values could be anything someone wants to log: metrics, learning rate, time, etc.

**Comment by kshen3778** [comment]
Yup, makes sense. Perhaps `log_values()` could take in a dictionary of values much like how a similar function handles it in PTL: https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/loggers/wandb.py#L213 ?

Also, when could we remove the old arguments such as `wandb` and `wandb_logger` for the model classes - is it only immediately before major releases such as 2.6.0? If so do we need to specify in the removal PR that this will cause some breaking changes and it should only be merged for the next major update?

**Comment by peastman** [comment]
At a minimum, I think we should go through a full major release in which it's deprecated but still present.  But there's also no particular need to remove it, since it adds minimal complexity.  A good general rule for deprecated features is to remove them only when it starts causing problems to keep them around.  There's no need to break compatibility without reason.

**Comment by kshen3778** [comment]
I see that makes sense. We can keep it in there, I was just worried there was going to be too many redundant arguments.

For functions such as log_score() and log_batch(), what would you suggest if the subclass were to override them and add extra arguments. For example, I'm currently imagining something like:

Logger:
```
def log_batch(self, step, loss, inputs, labels)

def log_values(self, step, values)
```
WandbLogger (subclass of Logger):
```
def log_batch(self, step, loss, inputs, labels, group)

def log_values(self, step, values, group, dataset_id):
```
There are specific arguments that would be passed into WandbLogger and useful only for Weights & Biases (such as dataset_id, group)

An example would be the extra argument `group` which is a string appended in front of the logged value's key in the Wandb dashboard to specify where it came from (for example appending "eval" to the roc_auc_score key to create "eval/roc_auc_score": this would create the [roc_auc_score chart under the category "eval"](https://wandb.ai/kshen/deepchem_graphconv/runs/w4fqxhwf?workspace=user-kshen)). 
Since KerasModel/ValidationCallback is only responsible for calculating the values, it only makes sense to just pass in the relevant values to `log_scores()` and `log_batch()` without extra modification to the key names. Any modification or data processing before logging would be handle by the logger subclass.

I'm unsure whether or not it is a good practice for the subclass methods to have a different method signature. If not, perhaps we would need to decide on what the arguments for each method will be? Or maybe we can use `**kwargs` so the signature looks like `log_values(self, step, values, **kwargs)`.

What are you thoughts on this?

**Comment by peastman** [comment]
> I'm unsure whether or not it is a good practice for the subclass methods to have a different method signature.

If you invoke the method using the signature defined by the parent class, it needs to work correctly.  So it's alright for the subclass to add arguments as long as 1) they come after all arguments defined by the parent class, and 2) they have default values so it still works if they're omitted.
